### ISSUE TITLE
Orchard Core intermittently displays login options when only one external provider is configured

### ISSUE THREAD

**Issue by dturska** [OrchardCMS/OrchardCore#17219]
### Describe the bug

When using Orchard Core with a single external identity provider configured for login and the option "Use external provider for login" enabled in the admin settings, the login screen intermittently displays a choice between the external provider and the standard username/password login. It is expected that the system should automatically redirect to the external provider for authentication without showing the login screen.

### Orchard Core version

2.1.2

### To Reproduce

1. Go to Settings > External Login in the Orchard Core admin dashboard.
2. Enable the "Use external provider for login" checkbox.
3. Ensure only one external provider is configured (e.g., Github).
4. Access the login page repeatedly using the same URL (e.g., `https://localhost:5001/admin`).

### Expected behavior

The login screen should consistently redirect to the external provider for authentication without presenting the choice between username/password login and the external provider.

### Actual behavior

The behavior is inconsistent and alternates between the following:
- On the first page load, it redirects to the external provider as expected.
- On subsequent page loads (odd vs. even attempts), the login screen displays a choice between the external provider and the standard login.

For example:
- First load of `https://localhost:5001/admin` works and redirects.
- Second load shows the login screen with options.
- Subsequent attempts alternate between the two behaviors.

### Steps to Mitigate/Workaround

No workaround identified.

### Suggested Fix

Update the logic in the login flow to ensure consistent behavior when the "Use external provider for login" option is enabled, particularly when only one external provider is configured. The system should always redirect to the external provider without showing the login screen.

### References

N/A

**Comment by hishamco** [comment]
It would be nice if we give the first-time authors to start 1st PR in OC

**Comment by MikeAlhayek** [comment]
> It would be nice if we give the first-time authors to start 1st PR in OC

Yes that would be super awesome! But, the author should show interest into providing a fix first :)

**Comment by dturska** [comment]
I am still encountering this issue in version 2.1.3. While I noticed that the update added code to the LoggedInAsync event, I am reproducing this error without logging in.

Scenario:
1) Without logging in, if I navigate to https://localhost/admin, it initially redirects as expected to the Identity Provider.

2) However, if I then paste https://localhost/admin into the browser and press Enter (without logging in), the OrchardCore login screen appears. It shows the local login option along with the external provider button.

3) Repeating this process alternates between redirecting to the external provider and showing the OrchardCore login screen with the local and external login options.

Observation:
The issue seems to stem from the public async Task<IActionResult> LoggingInAsync() method.

Workaround:
Locally, I’ve implemented a workaround by overriding the ExternalLoginFormEvents and clearing the tempData with the key ExternalLoginAutoRedirectKeyName every time. This doesn't seem like the proper fix though. 
 
 
![image](https://github.com/user-attachments/assets/fc0e54c0-28df-4f48-8607-0038ac78d8d5)

**Comment by MikeAlhayek** [comment]
You don't want to remove that key every time as it could cause an infinite loop for your users.

I think we need to change how we store that key to something that will be removed on the next request.

**Comment by asteekgoswami** [comment]
I'm experiencing the same issue  —/login inconsistently alternates between redirecting to the external provider and showing the local login form. Has this been fixed or is there a workaround available?

**Comment by MikeAlhayek** [comment]
@asteekgoswami which version are you using?

**Comment by asteekgoswami** [comment]
@MikeAlhayek  2.1.7

**Comment by okalangkenneth** [comment]
I would like to take a stab at this.
I think the root cause is that `TempData.ContainsKey()` in `ExternalLoginFormEvents.cs`, which marks the entry for deletion at the end of the request. After the redirect fires on the first request, the key is consumed on the second request's check, leaving the third request with no key. Thus, the alternating behavior.
The fix would be to replace `ContainsKey `with `Peek != null` on line 56, so the key persists until` LoggedInAsync` removes it on successful login. This also handles the external provider failure case correctly...if the provider fails and the user lands back on the login page, the key stays set and shows the form rather than looping. Happy to open a PR if this approach makes sense.

**Comment by gvkries** [comment]
We'd appreciate if you try to fix this issue. But if I understand this correctly, the fix you describe is not sufficient. In case of a single external identity provider, the login form should not be displayed at all. That must be considered here as well.

**Comment by okalangkenneth** [comment]
What should happen if the external provider fails, e.g., if the user cancels on the IdP side and is redirected back? Should the system still redirect to the provider again, or is showing the login form acceptable in that specific failure scenario? If I know this, I will keep the TempData with a smarter lifecycle or switch to a shorter-lived cookie cleared in the callback handler...or something completely different.

**Comment by gvkries** [comment]
IMHO a error message would be the best in this case. In any case, we must make sure there is no endless redirect loop in error cases. 
@MikeAlhayek can you help with designing this in a better way?

**Comment by MikeAlhayek** [comment]
I agree with @gvkries I would display an error message otherwise the user won't know why they keep getting redirected.

**Comment by okalangkenneth** [comment]
So here is my approach am thinking:
Remove the TempData guard completely from `LoggingInAsync()` as it is the source of the alternating bug. Then, in the external login callback error handler, when the IdP returns a failure, redirect back to the login page with an error query parameter like `?externalLoginError=true`. 

In `LoggingInAsync()`, check for that query parameter..if present, skip the auto-redirect and let the login form render with an error message explaining the external login failed. This way, the normal flow always redirects, failure flow shows the login form with an error message, and there is no infinite loop since the error parameter breaks the redirect chain.

Makes some sense? Happy to implement it if so.

**Comment by okalangkenneth** [comment]
I've implemented the approach. The changes are on my fork at:

https://github.com/okalangkenneth/OrchardCore/tree/fix/external-login-redirect-17219

Summary of changes:

- `ExternalLoginFormEvents.cs`. Added `externalLoginError` query parameter check before the auto-redirect, so the login form renders with the error message when returning from a failed IdP attempt
- `ExternalAuthenticationsController.cs`. `RedirectToLogin(string returnUrl)` now appends `externalLoginError=true` to all external login failure redirects

Happy to open a PR if the approach looks good.

**Comment by okalangkenneth** [comment]
Hi @hishamco . I notice the issue was closed as complete. Could you point me to the fix that was merged? The development section still links to #17220 from December 2024, which was the fix that @dturska and @asteekgoswami confirmed was broken in later versions. I have an implementation ready if it's still needed.

**Comment by hishamco** [comment]
The PR #17220 should fix the issue

**Comment by gvkries** [comment]
> The PR [#17220](https://github.com/OrchardCMS/OrchardCore/pull/17220) should fix the issue

This issue is not completely fixed.

**Comment by gvkries** [comment]
@okalangkenneth Please open a PR, there we can discuss this further.

**Comment by hishamco** [comment]
> This issue is not completely fixed.

Strange!! I thought it was completely fixed
